### ISSUE TITLE
Statuscode 404 from istio-proxy when cluster isn't found is missleading

### ISSUE THREAD

**Issue by Stono** [istio/istio#12760]
Hey, 
We've been going through the process of using `Sidecar` to scope all of our sidecars (which works beautifully I must add), however there is one behaviour that I think needs looking at.

We use both NetworkPolicy and Sidecar to restrict Egress.  Previously we only had NetworkPolicy, so when `appA` tried to talk to `appB` (when it shouldn't be able to), it'd network timeout and envoy would return a 503.  This wasn't brilliant, but it was quite obvious that there was a connection issue.

However, now that `Sidecar` doesn't include `appB`, envoy will immediately 404 the request, as the cluster isn't found.

This however, is miss-leading, and doesn't indicate to the sending app that there was actually anything wrong with the request, the sending app just believes the remote endpoint returned a 404.   This is a problem, say for example that remote endpoint was expected to return 404 for some operation like "checkUserExists".

I'm not sure what the appropriate way to handle this would be, perhaps `502 Bad Gateway` would be more appropriate.

PS.  This caused us a live issue today as a misconfigured sidecar policy broke the egress, and we didn't know as the source app just thought the destination was acceptably 404'ing.

PPS.  Are there any envoy metrics we could alert on that would signify this black hole (@douglas-reid?)?

**Comment by Stono** [comment]
or 418, naturally.

**Comment by Stono** [comment]
ping @costinm @rshriram @nmittler - this one is actually quite a pressing issue for us, would be great to get some input.

**Comment by louiscryan** [comment]
Yeah, that's definitely an issue. 

Even if we don't make 503 the default its certainly reasonable to allow control over the 'this service wasn't mapped in' status codes. 

A reasonable alternative to 503 would be '403 Forbidden', would be good to hear if that would cause application-level issues for you?. 

403 indicates that the failure state is permanent and not a function of a temporary network outage. Given that the configuration scoping is intentional you might argue that 403 is a better fit

**Comment by Stono** [comment]
Any 4x code is miss-leading in my opinion and isn't indicative of a gateway problem (which is what this is).

502 BAD GATEWAY is easily the best code for this.

This is that much of an issue that I need to build a monkeypatched version of `1.1.1` proxy to deal with it.  Can anyone point me to the code where the 404 is being returned from?

**Comment by louiscryan** [comment]
I'm not sure I entirely agree that 502 is universally the best option but it may be the most pragmatic.

In the meantime have you looked into defining your own 'black hole' virtualservice that sets

https://istio.io/docs/reference/config/networking/v1alpha3/virtual-service/#HTTPFaultInjection-Abort

The order of routing of the hosts referenced in a Sidecar is ordered so if you did something like...

VirtualService (namespace: blackhole, host: *, destination: <doesnt matter>, fault-inject: 503, percent: 1.0)

Note: that we namespace it so it doesn't get referenced by default. Then we do

Sidecar (egress: : [ real-host1, real-host2, blackhole/* ]) 

you should get the behavior you want as the routing rules will fall-through.

**Comment by Stono** [comment]
As it stands we have 190 Sidecars in isolated namespaces, it feels like a global problem for me than something that i need to configure on each sidecar.

**Comment by Stono** [comment]
@louiscryan any idea where i can find the 404 response in code?

**Comment by louiscryan** [comment]
Sure.  The config-gen is 

httproute.go:buildSidecarOutboundHTTPRouteConfig

there's an equivalent example for Gateway (not Sidecar) that is explicitly generating a wildcard match to serve 404 if no routes are defined 

gateway.go:buildGatewayHTTPRouteConfig ~line 286-293 (creates a DirectResponse) which is effectively what you'd want to do though it should be done before host merging occurs in 'mergeAllVirtualHosts'

HTH

@howardjohn

**Comment by Stono** [comment]
Thanks @louiscryan - i'll have a read and see if there is a pragmatic way i can patch this.

**Comment by howardjohn** [comment]
I want to make sure I fully understand this:

Say we add a wildcard match for each port as the (as the last virtual host) that gives a 502.

Any requests to this port that doesn't match will be get a 502, not just things that aren't imported. So `curl service.not-in-scope-namespace` will return 502, but also `curl 1.1.1.1` (or some random address). Is this what we are looking for?

We could make it only 502 for real services that are out of scope, but that defeats the main purpose of the Sidecar, to reduce config for scalability, so I don't think that is a good option.

@Stono

**Comment by Stono** [comment]
Hey,
Yes that's the behaviour i'd be expecting, going to `some-random-fqdn`, or `some-random-ip` should both be `502` gateway errors unless they're defined as services.

404 is majorly miss-leading and doesn't tell the sender that there was an issue talking to the remote service, it tells the sender that the remote service returned 404, which is not true.

**Comment by Stono** [comment]
Just to update this with out slack chat for clarity of others reading:

route not defined in Sidecar, `REGISTRY_ONLY` = gateway error (502)
route not defined in Sidecar, `ALLOW_ANY` = off to the service

https://istio.io/docs/tasks/traffic-management/egress/

**Comment by howardjohn** [comment]
Change here: https://github.com/istio/istio/pull/12916

@louiscryan I used 502 error as a placeholder, happy to change it though.

@costinm @rshriram this is a fairly substantial behavior change if you want to weigh in. It is blocked by a feature flag though.

**Comment by Stono** [comment]
We had another production outage caused by this today.  End to end tests didn't pick it up because 404 doesn't flag as an invalid response from the repo api.

@howardjohn will this be in 1.0.2?  If not, could you produce a 1.0.2 build with this patch so i can use it as our proxy image?  It's a real issue for us.

<img width="1373" alt="Screenshot 2019-04-03 at 10 42 55" src="https://user-images.githubusercontent.com/1486729/55469360-48530d00-55fd-11e9-83fd-cc32e832b600.png">

**Comment by Stono** [comment]
Using `pilot.env.PILOT_ENABLE_FALLTHROUGH_ROUTE=1` with the latest 1.1 nightly (`1.1-20190404-09-16`) now correctly gives 502s :-)

Thanks @howardjohn 

```
root@platform-testing-6f98ccd788-jg298:/app# curl -v http://app.moo
* Rebuilt URL to: http://app.moo/
*   Trying 10.192.98.243...
* TCP_NODELAY set
* Connected to app.moo (10.192.98.243) port 80 (#0)
> GET / HTTP/1.1
> Host: app.moo
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
< location: http://app.moo/
< date: Thu, 04 Apr 2019 14:44:35 GMT
< server: envoy
< content-length: 0
<
* Curl_http_done: called premature == 0
* Connection #0 to host app.moo left intact
```

**Comment by howardjohn** [comment]
Going to close this - will be available in 1.1.3 or using the nightly builds.
